### PR TITLE
Add extractors for 2m.ma

### PR DESCRIPTION
**Pull request by CrankDatSouljaBoy**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This adds an extractor for 2m.ma, the website for the biggest TV channel in Morocco.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
> * [x]  At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-29T05%3A20%3A16Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-29T05%3A20%3A16Z#yt-dlp-coding-conventions)

Read again

File: yt_dlp/extractor/deuxm.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            'ext': 'mp4',
+            'title': 'Journal  Amazigh : Vendredi 28 Octobre 2022',
+            'thumbnail': 'https://2msoread-ww.amagi.tv/mediasfiles/videos/images/2022/10/28/16669764903.png'
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        info = self._download_json(
+            'https://2m.ma/api/watchDetail/%s' % (video_id),
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:32]: fstring

File: yt_dlp/extractor/deuxm.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            video_id, "Downloading media JSON") or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:33]: single quotes

File: yt_dlp/extractor/deuxm.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        video = traverse_obj(info, ('response', 'News'))
+        title = video.get('titre')
+        thumbnail = url_or_none(video.get('image'))
+        return {
+            'id': video_id,
+            'title': title,
+            'url': video.get('url'),
+            'description': video.get('description'),
+            'thumbnail': thumbnail
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:42]: can be simplified

File: yt_dlp/extractor/deuxm.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:43]:
```suggestion
        video = self._download_json(
            f'https://2m.ma/api/watchDetail/{video_id}', video_id)['response']['News']
        return {
            'id': video_id,
            'title': video.get('titre'),
            'url': video['url'],
            'description': video.get('description'),
            'thumbnail': url_or_none(video.get('image')),
        }
```

File: yt_dlp/extractor/deuxm.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+class DeuxMNewsIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?2m\.ma/[^/]+/news/(?P<id>(?:.)+)'
+
+    _TESTS = [{
+        'url': 'https://2m.ma/fr/news/Sahara-la-r%C3%A9solution-2654-de-l-ONU-une-confirmation-de-l-appui-%C3%A0-l-initiative-d-autonomie-Vid%C3%A9o--20221028',
+        'info_dict': {
+            'id': '635c4aad346a31673b1aaec3',
+            'ext': 'mp4',
+            'title': 'Sahara : la r\u00e9solution 2654 de l\u0027ONU, une confirmation de l\u0027appui \u00e0 l\u0027initiative d\u0027autonomie (Vid\u00e9o)',
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:54]: use md5

File: yt_dlp/extractor/deuxm.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            'description': 'md5:a240fb360871e6ffd37e1fadd813e42c',
+            'thumbnail': 'https://2msoread-ww.amagi.tv/mediasfiles/videos/images/2022/10/28/16669845857-1.png'
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:56]: is this permanant? If not, use regex

File: yt_dlp/extractor/deuxm.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        'info_dict': {
+            'id': '635405760ed0533c940e7e82',
+            'ext': 'mp4',
+            'title': 'MD Talks: Des diplomates louent le potentiel \u00e9nerg\u00e9tique de la r\u00e9gion de Guelmim-Oued Noun ',
+            'thumbnail': 'https://2msoread-ww.amagi.tv/mediasfiles/videos/images/2022/10/22/16664504981-Reportage---MD-TALKS.png'
+        }
+    }]
+
+    def _real_extract(self, url):
+        lang = self._search_regex(r'https?://(?:www\.)?2m\.ma/(?P<lang>(?:[a-z])+)/.*', url, 'lang')
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:69]: capture this from _VALID_URL

File: yt_dlp/extractor/deuxm.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        article = traverse_obj(info, ('response', 'article'))[0]
+        title = article.get('title')
+        thumbnail = url_or_none(article.get('cover'))
+        return {
+            'id': article.get('id'),
+            'title': title,
+            'url': article.get('image')[0],
+            'description': article.get('content'),
+            'thumbnail': thumbnail
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:83]: same as above

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/deuxm.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
+from .common import InfoExtractor
+from ..utils import url_or_none
+
+
+class DeuxMIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?2m\.ma/[^/]+/replay/single/(?P<id>(?:[A-Za-z0-9_.]{1,24})+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:6]: What do u intend by `(?:[A-Za-z0-9_.]{1,24})+`? Did you mean just `[\w.]{1,24}`?

File: yt_dlp/extractor/deuxm.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            'ext': 'mp4',
+            'title': 'Interview: Casablanca hors des sentiers battus avec Abderrahim KASSOU (Replay) ',
+            'description': 'md5:3b8e78111de9fcc6ef7f7dd6cff2430c',
+            'thumbnail': 'https://2msoread-ww.amagi.tv/mediasfiles/videos/images/2022/10/17/1666031205e8b5613cf7fd24cc10a116844b23bfd4.png'
+        }
+    }]
+
+    def _real_extract(self, url):
+        lang = self._search_regex(self._VALID_URL, url, 'lang')
+        article_name = self._match_id(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:68]:
```suggestion
        article_name, lang = self._match_valid_url(url).group('id', 'lang')
```

File: yt_dlp/extractor/deuxm.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            'id': video_id,
+            'title': video.get('titre'),
+            'url': video['url'],
+            'description': video.get('description'),
+            'thumbnail': url_or_none(video.get('image')),
+        }
+
+
+class DeuxMNewsIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?2m\.ma/(?P<lang>(?:[a-z])+)/news/(?P<id>(?:.)+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:42]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?2m\.ma/(?P<lang>\w+)/news/(?P<id>[^/#?]+)'
```

File: yt_dlp/extractor/deuxm.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }]
+
+    def _real_extract(self, url):
+        lang = self._search_regex(self._VALID_URL, url, 'lang')
+        article_name = self._match_id(url)
+        video = self._download_json(
+            f'https://2m.ma/api/articlesByUrl?lang={lang}&url=/news/{article_name}', article_name)['response']['article'][0]
+        return {
+            'id': video.get('id'),
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:72]:
```suggestion
            'id': video['id'],
```
id is mandatory

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/deuxm.py (hunk 9 above)
Comment (CrankDatSouljaBoy) [reply yt_dlp/extractor/deuxm.py:6]: indeed, i do not have much knowledge of regex

File: yt_dlp/extractor/deuxm.py (hunk 11 above)
Comment (CrankDatSouljaBoy) [reply yt_dlp/extractor/deuxm.py:42]: above

File: yt_dlp/extractor/deuxm.py (hunk 6 above)
Comment (pukkandan) [reply yt_dlp/extractor/deuxm.py]: why was this marked as resolved, @CrankDatSouljaBoy

Comment (CrankDatSouljaBoy) [reply yt_dlp/extractor/deuxm.py]: @pukkandan, this was an error on my side, i have used a regex now

File: yt_dlp/extractor/deuxm.py [hunk 13]
```diff
@@ -7,12 +7,12 @@ class DeuxMIE(InfoExtractor):
 
     _TESTS = [{
         'url': 'https://2m.ma/fr/replay/single/6351d439b15e1a613b3debe8',
-        'md5': '7ab69a170493c7b7c3febe2ad965c9fd',
+        'md5': '5f761f04c9d686e553b685134dca5d32',
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py:10]: how come md5 changed?

File: yt_dlp/extractor/deuxm.py [hunk 14]
```diff
@@ -58,8 +57,7 @@ class DeuxMNewsIE(InfoExtractor):
             'id': '634d9e108b70d40bc51a844b',
             'ext': 'mp4',
             'title': 'Interview: Casablanca hors des sentiers battus avec Abderrahim KASSOU (Replay) ',
-            'description': 'md5:3b8e78111de9fcc6ef7f7dd6cff2430c',
-            'thumbnail': 'https://2msoread-ww.amagi.tv/mediasfiles/videos/images/2022/10/17/1666031205e8b5613cf7fd24cc10a116844b23bfd4.png'
```
Comment (pukkandan) [inline yt_dlp/extractor/deuxm.py]: why is this removed?

File: yt_dlp/extractor/deuxm.py (hunk 13 above)
Comment (CrankDatSouljaBoy) [reply yt_dlp/extractor/deuxm.py:10]: i mishashed it earlier

File: yt_dlp/extractor/deuxm.py (hunk 14 above)
Comment (CrankDatSouljaBoy) [reply yt_dlp/extractor/deuxm.py]: should i add the regex to every thumbnail test?

**Comment by pukkandan** [comment]
Have you actually run the tests and verified it passes?

**Comment by CrankDatSouljaBoy** [comment]
> Have you actually run the tests and verified it passes?

how I figured I had mishashed it earlier..
